### PR TITLE
DevTools scheduling profiler: Add React component measures

### PR DESCRIPTION
**Pull request by bvaughn**
- [x] Add new user timing API marks for when components start/stop rendering to the scheduling profiler data set.
- [x] Add new row of React components (user code). These are technically visible in the flame graph, but they're too buried. I think it would be helpful to show which components are being rendered (and when) on a flattened row.
- [x] Added borders/bars to Flamechart view
- [x] Update record/cancel buttons to account for scheduling profiler tab

Add new row of React components (user code).

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js [hunk 1]
```diff
@@ -958,6 +968,9 @@ function updateFunctionComponent(
 
   let nextChildren;
   prepareToReadContext(workInProgress, renderLanes);
+  if (enableSchedulingProfiler) {
+    markComponentRenderStarted(workInProgress);
```
Comment (jstejada) [inline packages/react-reconciler/src/ReactFiberBeginWork.new.js:972]: so, you basically already know all the places where "render starts" and "render stops", and we're adding them all over, right?

File: packages/react-reconciler/src/ReactFiberThrow.new.js [hunk 2]
```diff
@@ -244,6 +247,7 @@ function throwException(
     }
 
     if (enableSchedulingProfiler) {
+      markComponentRenderStopped();
```
Comment (jstejada) [inline packages/react-reconciler/src/ReactFiberThrow.new.js:250]: so suspending is considered as render stopped, at least for the purposes of what we show in dev tools?

File: packages/react-reconciler/src/__tests__/SchedulingProfiler-test.internal.js [hunk 3]
```diff
@@ -539,6 +563,8 @@ describe('SchedulingProfiler', () => {
       )}-Example`,
       '--passive-effects-stop',
       `--render-start-${formatLanes(ReactFiberLane.DefaultLane)}`,
+      '--component-render-start-Example',
+      '--component-render-stop',
```
Comment (jstejada) [inline packages/react-reconciler/src/__tests__/SchedulingProfiler-test.internal.js:567]: cool!

File: packages/react-devtools-shared/src/devtools/views/Profiler/ClearProfilingDataButton.js [hunk 4]
```diff
@@ -8,22 +8,41 @@
...
+  let doesHaveData = false;
+  if (selectedTabID === 'scheduling-profiler') {
+    doesHaveData = schedulingProfilerData !== null;
+  } else {
+    doesHaveData = didRecordCommits;
+  }
+
+  const clear = () => {
+    if (selectedTabID === 'scheduling-profiler') {
+      clearSchedulingProfilerData();
```
Comment (jstejada) [inline packages/react-devtools-shared/src/devtools/views/Profiler/ClearProfilingDataButton.js:37]: could you share a bit more context on this change? what's the difference between clearing the scheduling profiler data vs clearing the profiler store?

the profiler store corresponds to the current profiler tab vs the new scheduling profiler tab?

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js [hunk 5]
```diff
@@ -357,6 +361,9 @@ function updateForwardRef(
   // The rest is a fork of updateFunctionComponent
   let nextChildren;
   prepareToReadContext(workInProgress, renderLanes);
+  if (enableSchedulingProfiler) {
```
[the remark this answers is not in the document]
Comment (bvaughn) [reply packages/react-reconciler/src/ReactFiberBeginWork.new.js:364]: Yeah, we double wrap things like this 😄

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js (hunk 1 above)
Comment (bvaughn) [reply packages/react-reconciler/src/ReactFiberBeginWork.new.js:972]: Yes.

File: packages/react-reconciler/src/ReactFiberThrow.new.js (hunk 2 above)
Comment (bvaughn) [reply packages/react-reconciler/src/ReactFiberThrow.new.js:250]: Yup!

File: packages/react-devtools-shared/src/devtools/views/Profiler/ClearProfilingDataButton.js (hunk 4 above)
Comment (bvaughn) [reply packages/react-devtools-shared/src/devtools/views/Profiler/ClearProfilingDataButton.js:37]: Yes. There are two profilers, each with a separate data source. One is React-generated and the other comes from the browser's Performance profile JSON. This kind of sucks, and I'd like to smooth over this distinction in the future (see #22015).

**Comment by mondaychen** [comment]
Looks like "reload and profile" button is always disabled for "scheduling profiler" (now called "timeline") by purpose. What's the reason? It recently reported as a bug #24603

**Comment by lunaruan** [comment]
@mondaychen There's no reason. It's most likely a left over artifact from when the only way to use Timeline Profiler was to upload a chrome profiling trace.

**Comment by mondaychen** [comment]
@lunaruan Thanks. I'll remove that restriction then.

**Pull request merged**